src/main/java/com/arpnetworking/utility/RegexAndMapReplacer.java

```diff
@@ -0,0 +1,165 @@
+ * limitations under the License.
+ */
+package com.arpnetworking.utility;
+
+import java.util.Map;
+import java.util.regex.Matcher;
+
+/**
+ * A regex replacement utility that can also replace tokens not found in the regex.
+ *
+ * @author Brandon Arp (brandon dot arp at inscopemetrics dot com)
+ */
+public final class RegexAndMapReplacer {
+    /**
+     * Replaces all instances of $n (where n is 0-9) with regex match groups, ${var} with regex capture group or variable (from map) 'var'.
+     *
+     * @param matcher matcher to use
+     * @param input input string to match against
+     * @param replace replacement string
+     * @param variables map of variables to include
+     * @return a string with replacement tokens replaced
+     */
+    public static String replaceAll(final Matcher matcher, final String input, final String replace, final Map<String, String> variables) {
+
+        matcher.reset();
+        boolean found = matcher.find();
+        if (found) {
+            final StringBuilder builder = new StringBuilder();
+            int lastMatchedIndex = 0;
+            do {
+                builder.append(input.substring(lastMatchedIndex, matcher.start()));
+                lastMatchedIndex = matcher.end();
+                appendReplacement(matcher, replace, builder, variables);
+                found = matcher.find();
+            } while (found);
+            // Append left-over string after the matches
+            if (lastMatchedIndex < input.length() - 1) {
+                builder.append(input.substring(lastMatchedIndex, input.length()));
+            }
+            return builder.toString();
+        }
+        return input;
+    }
+
+    private static void appendReplacement(final Matcher matcher, final String replacement, final StringBuilder replacementBuilder,
+                                          final Map<String, String> variables) {
+        final StringBuilder tokenBuilder = new StringBuilder();
+        int x = -1;
+        while (x < replacement.length() - 1) {
+            x++;
+            final char c = replacement.charAt(x);
+            if (c == '\\') {
+                x++;
+                processEscapedCharacter(replacement, x, replacementBuilder);
+            } else {
+                if (c == '$') {
+                    x += writeReplacementToken(replacement, x, replacementBuilder, matcher, variables, tokenBuilder);
+                } else {
+                    replacementBuilder.append(c);
+                }
+            }
+        }
+    }
+
+    private static void processEscapedCharacter(final String replacement, final int x, final StringBuilder builder) {
+        if (x >= replacement.length()) {
+            throw new IllegalArgumentException(
+                    String.format("Improper escaping in replacement, must not have trailing '\\' at col %d: %s", x, replacement));
+        }
+        final Character c = replacement.charAt(x);
+        if (c == '\\' || c == '$' || c == '}') {
```

Comment:
I believe with most open-close style formats you need to escape both. I understand it's strictly not necessary here, but I'd like to be consistent with say regex and bash.